### PR TITLE
storage: be tighter with allocations when TargetBytes is present

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, while `put`ting the key into the `repr`, we could make an
allocation that was too large given the remaining `TargetBytes` budget.
This is the case since we're exponentially increasing the capacities of
the buffers until 128MiB and because we're only accounting for the
length of the slice even though the whole capacity would have a memory
footprint.

For example, with 10MiB of `TargetBytes` (which is used by SQL in many
cases) and a ScanResponse that exceeds that limit, we'd allocate
capacities that are powers of two, starting at, say, 256B, and would go
all the way up to 8MiB; however, given that 10MiB limit, we'd only use
2MiB of that last 8MiB `repr` when we encounter the target bytes limit
and stop. Such behavior is kinda ok if the response is marshalled by the
gRPC to be sent to the other node, but it is quite unfortunate in the
local fast-path cases (meaning the SQL and the KV are part of the same
physical machine). In the latter scenario SQL would only account for the
lengths of slices while keeping the whole slices alive for a while,
leading to significant unaccounted for memory. In the example above, on
the order of 6MiB would be unaccounted for - multiply that by some
concurrency, and we have unaccounted memory on the order of hundreds of
MiBs.

The current behavior seems especially bad for the streamer use case
where we issue many requests with the `TargetBytes` set and use
`ScanResponse.NumBytes` field (which tracks the lengths of the slices)
for the memory accounting purposes.

In order to improve here, this commit teaches `put` method about the
maximum capacity it can use. In the example above, the last slice would
be on the order of 2MiB making everything happy: we stay very close to
TargetBytes limit and don't have any wasted space.

Addresses: #64906.
Addresses: #82160.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
An example of the current behavior, only 10MiB are accounted for whereas the actual footprint is 16MiB:
Screen Shot 2022-07-12 at 1 46 55 PM

**Comment by yuzefovich** [comment]
I'm curious to hear what people think and very open to other solutions, but I believe we should fix the oversight in one way or another.

**Comment by yuzefovich** [comment]
Pulled the first commit into a separate PR so that we can focus here on the main change. PTAL.

**Review by erikgrinaker — APPROVED** [review]
Thanks! This is a nice and simple improvement, but I think it can still underestimate the memory use quite significantly in pathological cases where we're only using half of each buffer. Ideally, I suppose we'd want these to form a continuous buffer where KVs can straddle the bounds to avoid any wasted memory. Maybe write up an issue if that's something we want to pursue.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @sumeerbhola, and @yuzefovich)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1032 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7GyJB23La_t3uyfFd1:-N7GyJB23La_t3uyfFd2:b-ib7d2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd1564dff0c19493ebaf8cec0c1f547db9691716/pkg/storage/pebble_mvcc_scanner.go#L1032)):*
> ```Go
> 		}
> 	}
> ```

There is a bunch of targetBytes and maxKeys logic in this function that is making it harder to understand. How about a comment like the following

// We are here due to one of the following cases:
// A. No limits were exceeded
// B. Limits were exceeded, but we need to put a key, so mustPutKey = true
// For B we will never set maxNewSize.
// For A, we may set maxNewSize, but we already know that p.targetBytes > p.results.bytes. So maxNewSize > 0. However, maxNewSize may not be enough to hold the result, in which case added can be false.

Don't we have a bug in this new logic in that case? In that we will call `keyNotAdded` below without doing the `mustPutKey` dance, so we could remove everything even if `!p.allowEmpty`?
If we lifted the following logic above the code that handles `p.resumeReason != 0` then it seems correct:
```
if p.resumeReason == 0 {
	var maxNewSize int
	if p.targetBytes > 0 && p.targetBytes > p.results.bytes {
		maxNewSize = int(p.targetBytes - p.results.bytes)
	}
	if added, err := p.results.put(ctx, rawKey, rawValue, p.memAccount, maxNewSize); err != nil {
		p.err = errors.Wrapf(err, "scan with start key %s", p.start)
		return false
	}
        if added {
           p.advanceKey()
        } else {
           p.resumeReason = roachpb.RESUME_BYTE_LIMIT
        }
}
// INVARIANT: p.resumeReason != 0.
```

Also the interplay between `p.targetBytesAvoidExcess` and the `maxNewSize` is not very clear. It seems we are ok with exceeding the size when `!p.targetBytesAvoidExcess`, but then this put allocation behavior may prevent an excess. Am I understanding correctly? A code comment would help.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1032 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7GyJB23La_t3uyfFd1:-N7H38ln6QXL0IvXfo2V:b-xed4vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd1564dff0c19493ebaf8cec0c1f547db9691716/pkg/storage/pebble_mvcc_scanner.go#L1032)):*
> // A. No limits were exceeded

This is not right - we are in this branch only if `p.resumeReason != 0` meaning that the limits were exceeded.

We only proceed to try to `put()` the current key while restricting the new capacity only if the limits were not exceeded. This actually means that adding the current key into the result won't put us over `targetBytes` limit, so we don't have to worry about `!p.allowEmpty` case - the `put` will succeed. This is ensured by https://github.com/cockroachdb/cockroach/blob/a8568d230568817a661c6eefbacd4e23feccdcd3/pkg/storage/pebble_mvcc_scanner.go#L976

I added some more commentary, let me know what you think.

> Also the interplay between p.targetBytesAvoidExcess and the maxNewSize is not very clear. It seems we are ok with exceeding the size when !p.targetBytesAvoidExcess, but then this put allocation behavior may prevent an excess. Am I understanding correctly? A code comment would help.

You're right that `!p.targetBytesAvoidExcess` case needs some attention. However, we are removing this field in #84519 now because `p.targetBytesAvoidExcees` is always `true` on 22.2 version (it was only kept around for 22.1-22.2 mixed version cluster setup), so we can actually ignore `!p.targetBytesAvoidExcess` case. I initially included the corresponding commit into this PR but decided to pull it out.

**Review by yuzefovich — COMMENTED** [review]
Good point. Filed #84594 to track that, but it seems such a scenario should occur rarely in practice, so I'm ok with that being addressed (if ever) later.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-:-N7HEdnK7RjyQsrRThq8:b-ckvamj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1032 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7GyJB23La_t3uyfFd1:-N7HCnn9DAG7hA5QMSdL:bs6g7k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd1564dff0c19493ebaf8cec0c1f547db9691716/pkg/storage/pebble_mvcc_scanner.go#L1032)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

> // A. No limits were exceeded

This is not right - we are in this branch only if `p.resumeReason != 0` meaning that the limits were exceeded.

We only proceed to try to `put()` the current key while restricting the new capacity only if the limits were not exceeded. This actually means that adding the current key into the result won't put us over `targetBytes` limit, so we don't have to worry about `!p.allowEmpty` case - the `put` will succeed. This is ensured by https://github.com/cockroachdb/cockroach/blob/a8568d230568817a661c6eefbacd4e23feccdcd3/pkg/storage/pebble_mvcc_scanner.go#L976

I added some more commentary, let me know what you think.

> Also the interplay between p.targetBytesAvoidExcess and the maxNewSize is not very clear. It seems we are ok with exceeding the size when !p.targetBytesAvoidExcess, but then this put allocation behavior may prevent an excess. Am I understanding correctly? A code comment would help.

You're right that `!p.targetBytesAvoidExcess` case needs some attention. However, we are removing this field in #84519 now because `p.targetBytesAvoidExcees` is always `true` on 22.2 version (it was only kept around for 22.1-22.2 mixed version cluster setup), so we can actually ignore `!p.targetBytesAvoidExcess` case. I initially included the corresponding commit into this PR but decided to pull it out.
</blockquote></details>

I temporarily cherry-picked the commit that removes `targetBytesAvoidExcess` logic into this PR to ease the discussion. Also note that the existing tests were, in fact, failing for `!p.targetBytesAvoidExcess` case.

**Comment by yuzefovich** [comment]
Rebased on top of latest master which contains the removal of `targetBytesAvoidExcess` business.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @sumeerbhola, and @yuzefovich)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1032 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7GyJB23La_t3uyfFd1:-N7LHEO0DXL4h1cV06xm:b-voxeef) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd1564dff0c19493ebaf8cec0c1f547db9691716/pkg/storage/pebble_mvcc_scanner.go#L1032)):*
> This is not right - we are in this branch only if p.resumeReason != 0 meaning that the limits were exceeded.

I was referring to the code after the if-block, i.e., a comment that would precede all of the code after the if-block.

> We only proceed to try to put() the current key while restricting the new capacity only if the limits were not exceeded. This actually means that adding the current key into the result won't put us over targetBytes limit, so we don't have to worry about !p.allowEmpty case - the put will succeed. This is ensured by ...

So we know `maxNewSize` is sufficient because  `lenToAdd := p.sizeOf(lenKey, lenValue)` used by `put` is the same computation used in in the avoid-excess case of `p.results.sizeOf(len(rawKey), len(rawValue)`. Relying on such matching code makes me slightly nervous -- there should be at least some invariant checking to ensure that the code in the 2 places doesn't slightly diverge so that the spare capacity in the former doesn't appear slightly larger than the latter and cause the latter (`put`) to fail.

```
	if p.targetBytes > 0 && (p.results.bytes >= p.targetBytes || (p.targetBytesAvoidExcess &&
		p.results.bytes+int64(p.results.sizeOf(len(rawKey), len(rawValue))) > p.targetBytes)) {
```
> You're right that !p.targetBytesAvoidExcess case needs some attention. ...
> Also note that the existing tests were, in fact, failing for !p.targetBytesAvoidExcess case.

That's the one where I thought there was a bug since maxNewSize was not guaranteed to hold the key-value pair. So glad to hear that our existing tests covered that case and were failing.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @sumeerbhola, and @yuzefovich)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 131 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LRMML9CnzivMG0Otx:-N7LRMML9CnzivMG0Oty:b-8y62hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L131)):*
> ```Go
> 				newSize *= 2
> 			}
> 		}
> ```

The old size code was hard to follow for me, and now it is even harder. How about something like:
```
// Exponential increase by default, while ensuring that we
// respect a hard lower bound of lenToAdd and the soft upper bound of
// maxSize and the hard upper bound of maxNewSize (if set).
if maxNewSize > 0 && maxNewSize < lenToAdd {
  // Hard upper bound is greater than hard lower bound, so cannot add.
  return false, nil
}
// Exponential growth to ensure newSize >= lenToAdd.
newSize := 2 * cap(p.repr)
if newSize == 0 || newSize > maxSize {
  // If the previous buffer exceeded maxSize, we don't double its capacity
  // for next allocation, and instead reset the exponential increase, in
  // case we had a stray huge key-value.
  newSize = minSize
}
for newSize < lenToAdd {
  newSize *= 2
}
// Respect soft upper-bound before hard lower-bound, since it could be lower than hard lower-bound.
if newSize > maxSize {
  newSize = maxSize
}
// Respect hard upper-bound.
if maxNewSize > 0 && newSize > maxNewSize {
  newSize = maxNewSize
}
// Now respect hard lower-bound.
if newSize < lenToAdd {
  newSize = lenToAdd
}

```
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1047 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LSfqoAseMab2XZIw5:-N7LSfqoAseMab2XZIw6:bifyggv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L1047)):*
> ```Go
> 
> 	var maxNewSize int
> 	if p.targetBytes > 0 && p.targetBytes > p.results.bytes && !mustPutKey {
> ```

// INVARIANT: !mustPutKey => maxNewSize is sufficient for key-value pair.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1053 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LT4jUB3KFXWUQ-dm_:-N7LT4jUB3KFXWUQ-dma:b8s7hl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L1053)):*
> ```Go
> 		p.err = errors.Wrapf(err, "scan with start key %s", p.start)
> 		return false
> 	} else if !added {
> ```

isn't this `!added` case a failure of an internal code invariant?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1032 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7GyJB23La_t3uyfFd1:-N7MQ55oDd9n3L9q1yNA:bxxbhxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd1564dff0c19493ebaf8cec0c1f547db9691716/pkg/storage/pebble_mvcc_scanner.go#L1032)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> This is not right - we are in this branch only if p.resumeReason != 0 meaning that the limits were exceeded.

I was referring to the code after the if-block, i.e., a comment that would precede all of the code after the if-block.

> We only proceed to try to put() the current key while restricting the new capacity only if the limits were not exceeded. This actually means that adding the current key into the result won't put us over targetBytes limit, so we don't have to worry about !p.allowEmpty case - the put will succeed. This is ensured by ...

So we know `maxNewSize` is sufficient because  `lenToAdd := p.sizeOf(lenKey, lenValue)` used by `put` is the same computation used in in the avoid-excess case of `p.results.sizeOf(len(rawKey), len(rawValue)`. Relying on such matching code makes me slightly nervous -- there should be at least some invariant checking to ensure that the code in the 2 places doesn't slightly diverge so that the spare capacity in the former doesn't appear slightly larger than the latter and cause the latter (`put`) to fail.

```
	if p.targetBytes > 0 && (p.results.bytes >= p.targetBytes || (p.targetBytesAvoidExcess &&
		p.results.bytes+int64(p.results.sizeOf(len(rawKey), len(rawValue))) > p.targetBytes)) {
```
> You're right that !p.targetBytesAvoidExcess case needs some attention. ...
> Also note that the existing tests were, in fact, failing for !p.targetBytesAvoidExcess case.

That's the one where I thought there was a bug since maxNewSize was not guaranteed to hold the key-value pair. So glad to hear that our existing tests covered that case and were failing.
</blockquote></details>

I see. I added the comments you were suggesting earlier given that we don't need to return that `added` boolean after all.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 131 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LRMML9CnzivMG0Otx:-N7MHKNL1WhAJskJRaV0:b-p1fz7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L131)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The old size code was hard to follow for me, and now it is even harder. How about something like:
```
// Exponential increase by default, while ensuring that we
// respect a hard lower bound of lenToAdd and the soft upper bound of
// maxSize and the hard upper bound of maxNewSize (if set).
if maxNewSize > 0 && maxNewSize < lenToAdd {
  // Hard upper bound is greater than hard lower bound, so cannot add.
  return false, nil
}
// Exponential growth to ensure newSize >= lenToAdd.
newSize := 2 * cap(p.repr)
if newSize == 0 || newSize > maxSize {
  // If the previous buffer exceeded maxSize, we don't double its capacity
  // for next allocation, and instead reset the exponential increase, in
  // case we had a stray huge key-value.
  newSize = minSize
}
for newSize < lenToAdd {
  newSize *= 2
}
// Respect soft upper-bound before hard lower-bound, since it could be lower than hard lower-bound.
if newSize > maxSize {
  newSize = maxSize
}
// Respect hard upper-bound.
if maxNewSize > 0 && newSize > maxNewSize {
  newSize = maxNewSize
}
// Now respect hard lower-bound.
if newSize < lenToAdd {
  newSize = lenToAdd
}

```
</blockquote></details>

I like it, thanks, done.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1047 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LSfqoAseMab2XZIw5:-N7MJNSJ0tgFaASmMhQ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L1047)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

// INVARIANT: !mustPutKey => maxNewSize is sufficient for key-value pair.
</blockquote></details>

Done.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1053 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-N7LT4jUB3KFXWUQ-dm_:-N7MOH0A8eOttUGl62ri:buj9kio) ([raw file](https://github.com/cockroachdb/cockroach/blob/2056a913b1b7b97885560cdfe3a4fe58b18ec8c4/pkg/storage/pebble_mvcc_scanner.go#L1053)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

isn't this `!added` case a failure of an internal code invariant?
</blockquote></details>

Yes, you're totally right, removed.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84319#-:-N7aFx3v0HxHX4MH4IRy:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

**Comment by yuzefovich** [comment]
Adjusted the commit message a bit.

TFTRs!

bors r+
